### ISSUE TITLE
new rule - ban default values on overload signatures

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ruff#23064]
### Summary

```py
from typing import overload

@overload
def foo(value: int = 1) -> int: ... # error: default value is useless. use ... instead

@overload
def foo(value: str) -> str: ...

def foo(value: int | str = 0) -> int | str: ...
```

specifying default values on overloads is bad because it's very easy to update the value in the implementation signature and forget to change it on all the overloads, which results in confusing signatures like in this example

**Comment by KotlinIsland** [comment]
if it could be validated to match the impl then it can be useful. what if the rule instead reports "default different in impl"?

**Comment by DetachHead** [comment]
that would still be duplicating the value for no reason. i think it's better to just enforce always using an ellipsis instead

**Comment by AlexWaygood** [comment]
A language server might in certain contexts filter out certain overloads (and the implementation) and only show relevant overloads when you hover over a function call, for example. Implementing this rule as suggested in the OP would hurt that use case. I think @KotlinIsland's suggested variation in https://github.com/astral-sh/ruff/issues/23064#issuecomment-3844941012 could be useful, though

**Comment by DetachHead** [comment]
imo the language server should just fill in the `...` on the resolved overload with the value from the implementation signature. perhaps it could also show them as an inlay hint

**Comment by AlexWaygood** [comment]
Sure, that's possible. It seems like enforcing this would just make things more complicated for those language servers for no obvious reason, though, given that we could just enforce that defaults in overloads should be consistent with the implementation — which would seem to also solve the problem your proposed rule seeks to solve, but would not make life more complicated for language servers in any way

**Comment by KotlinIsland** [comment]
> enforce that defaults match

kinda seems like busy work for the author that involves code duplication. I'd prefer detach's approach as it's more automated

maybe if some popular python LSP were to implement this feature then your suggestion might be more palatable (wink wink, nudge nudge)

**Comment by DetachHead** [comment]
as the maintainer of a python language server, i'm happy for this to be the responsibility of the language server. i don't think it really adds much complexity as it was only a couple of lines of code. see https://github.com/DetachHead/basedpyright/pull/1717

**Comment by Avasam** [comment]
From my time writing stubs in typeshed, including the default has quite often been beneficial to not make a mistake on the overloaded parameters (since if the type doesn't match your default, you'll get a type error). Introducing default values has also spotted existing mistakes. (most of the time it's about a `= ...` being left in whilst it should be impossible to match)

```py
@overload
def foo(a: str, b: Literal[False] = ...) -> None: ... # incorrect but no type error
@overload
def foo(a: int, b: Literal[True] = ...) -> None: ...
def foo(a: str | int, b: bool = True) -> None: ...
```
```py
@overload
def foo(a: str, b: Literal[False] = True) -> None: ... # aha ! Type error!
@overload
def foo(a: int, b: Literal[True] = True) -> None: ...
def foo(a: str | int, b: bool = True) -> None: ...
```
```py
@overload
def foo(a: str, b: Literal[False] = False) -> None: ... # still incorrect, no type error but Alex's suggestion of enforcing would fix this
@overload
def foo(a: int, b: Literal[True] = True) -> None: ...
def foo(a: str | int, b: bool = True) -> None: ...
```
```py
@overload
def foo(a: str, b: Literal[False]) -> None: ... # This is the correct signature
@overload
def foo(a: int, b: Literal[True] = True) -> None: ...
def foo(a: str | int, b: bool = True) -> None: ...
```


Granted it's easier for typeshed as it's all stubs-only, and we have tools to stay in sync with the implementation when it's in pure-python (stubtest)

**Comment by Avasam** [comment]
I'm over simplifying here, this may not be the best example off the top of my head, but I've also seen cases of using default values in an overload to communicate which default value could be used in a kwargs:
```py
@overload
def foo(*, a: str, bar: str = "str") -> str: ...
@overload
def foo(*, a: int, bar: str = "int") -> str: ...
def foo(**kwargs) -> str:
    # pretend something more complex is happening here that would justify the above signature
    return kwargs.setdefault("bar", type(kwargs["a"]).__name__)
```

**Comment by DetachHead** [comment]
> ```py
> @overload
> def foo(a: str, b: Literal[False] = ...) -> None: ... # incorrect but no type error
> @overload
> def foo(a: int, b: Literal[True] = ...) -> None: ...
> def foo(a: str | int, b: bool = True) -> None: ...
> ```

> ```py
> @overload
> def foo(a: str, b: Literal[False] = False) -> None: ... # still incorrect, no type error but Alex's suggestion of enforcing would fix this
> @overload
> def foo(a: int, b: Literal[True] = True) -> None: ...
> def foo(a: str | int, b: bool = True) -> None: ...
> ```

these *both* should be type errors. i don't see why the default value being ommitted from the overload should make any difference. i've opened https://github.com/astral-sh/ty/issues/2765 for this.

> I'm over simplifying here, this may not be the best example off the top of my head, but I've also seen cases of using default values in an overload to communicate which default value could be used in a kwargs:
> ```py
> @overload
> def foo(*, a: str, bar: str = "str") -> str: ...
> @overload
> def foo(*, a: int, bar: str = "int") -> str: ...
> def foo(**kwargs) -> str:
>     # pretend something more complex is happening here that would justify the above signature
>     return kwargs.setdefault("bar", type(kwargs["a"]).__name__)
> ```

sorry i don't understand this example at all lol

---

i guess this could be configurable if people disagree, but the main reason i think using ellipsis should be the preferred pattern is basically just [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself). even if a linter/type checker was able to correctly identify when one of the values is wrong, i would much rather only have to define the value in one place instead of havinf to maintain multiple duplicates of it.

**Comment by KotlinIsland** [comment]
> I'm over simplifying here, this may not be the best example off the top of my head, but I've also seen cases of using default values in an overload to communicate which default value could be used in a kwargs:
> 
> @overload
> def foo(*, a: str, bar: str = "str") -> str: ...
> @overload
> def foo(*, a: int, bar: str = "int") -> str: ...
> def foo(**kwargs) -> str:
>     # pretend something more complex is happening here that would justify the above signature
>     return kwargs.setdefault("bar", type(kwargs["a"]).__name__)

@Avasam 
```py
foo(a=True)
```

seems pretty fragile, more like abusing type annotations, this information should just go in the docstring
